File: Lowest Common Ancestor of a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # Time Complexity: O(n) n is the total number of elements
    # Space Complexity: O(h) h being the height of the tree
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root.val > p.val and root.val > q.val: return self.lowestCommonAncestor(root.left, p, q)
        elif root.val < p.val and root.val < q.val: return self.lowestCommonAncestor(root.right, p, q)
        else:
            return root

File: lowest-common-ancestor-of-a-binary-search-tree.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */

// Iterative
var lowestCommonAncestor = function(root, p, q) {
    if (!root) return new TreeNode();
    
    while (root != null) {
        if (p.val > root.val && q.val > root.val) root = root.right;
        else if (p.val < root.val && q.val < root.val) root = root.left;
        else return root;
    }
};

// Recursive
// var lowestCommonAncestor = function(root, p, q) {
//     if (!root) return new TreeNode();

//     if (p.val < root.val && q.val < root.val) {
//         return lowestCommonAncestor(root.left, p, q);
//     } else if (p.val > root.val && q.val > root.val) {
//         return lowestCommonAncestor(root.right, p, q);
//     } else {
//         return root;
//     }
// };
